packages/eslint-plugin/src/rules/no-redundant-type-constituents.ts

```diff
@@ -202,83 +308,118 @@ export default createRule({
-      typeNode: TSESTree.TypeNode,
-    ): TypeFlagsWithName[] {
-      const keywordTypeFlags = keywordNodeTypesToTsTypes.get(typeNode.type);
-      if (keywordTypeFlags) {
-        return [
-          {
-            typeFlags: keywordTypeFlags,
-            typeName: describeLiteralTypeNode(typeNode),
-          },
-        ];
+    const checker = services.program.getTypeChecker();
+
+    function reportRedundantTypes(
+      redundantTypes: Map<TSESTree.TypeNode, TypeRedundancyRelation[]>,
+      container: 'intersection' | 'union',
+    ) {
+      for (const [typeNode, typeRelations] of redundantTypes) {
+        const groupTypeRelationsByNonRedundantType =
+          getGroupTypeRelationsByNonRedundantType(typeRelations);
+
+        for (const [
+          nonRedundantType,
+          typeRelationAndNames,
+        ] of groupTypeRelationsByNonRedundantType) {
+          const nonRedundantTypeName = checker.typeToString(nonRedundantType);
+          const mergedRedundantTypeName = mergeTypeNames(
+            typeRelationAndNames.map(
+              ({ redundantTypeWithName }) => redundantTypeWithName,
+            ),
+            container === 'union' ? '|' : '&',
+          );
+
+          context.report({
+            node: typeNode,
+            messageId: 'typeOverridden',
+            data: {
+              container,
+              nonRedundantType: nonRedundantTypeName,
+              redundantType: mergedRedundantTypeName,
+            },
+          });
+        }
       }
+    }
 
+    function getUnionTypePart(
+      typeNode: ts.Node,
+      checker: ts.TypeChecker,
+    ): TypeWithName[] {
+      if (ts.isParenthesizedTypeNode(typeNode)) {
+        return getUnionTypePart(typeNode.type, checker);
+      }
+      if (ts.isUnionTypeNode(typeNode)) {
+        return typeNode.types.flatMap(typeNode =>
+          getUnionTypePart(typeNode, checker),
+        );
+      }
+      const type = checker.getTypeAtLocation(typeNode);
       if (
-        typeNode.type === AST_NODE_TYPES.TSLiteralType &&
-        typeNode.literal.type === AST_NODE_TYPES.Literal
+        ts.isTypeReferenceNode(typeNode) &&
+        tsutils.isUnionOrIntersectionType(type)
       ) {
-        return [
-          {
-            typeFlags:
-              primitiveTypeFlagTypes[
-                typeof typeNode.literal
-                  .value as keyof typeof primitiveTypeFlagTypes
-              ],
-            typeName: describeLiteralTypeNode(typeNode),
-          },
-        ];
+        const node = getTypeNodeFromReferenceType(type);
+        if (node && node !== typeNode) {
+          return getUnionTypePart(node, checker);
+        }
       }
+      return [
+        {
+          type,
+          typeName: checker.typeToString(type),
+        },
+      ];
+    }
 
-      if (typeNode.type === AST_NODE_TYPES.TSUnionType) {
-        return typeNode.types.flatMap(getTypeNodeTypePartFlags);
+    function getInterSectionTypePart(
```

Comment:
[Naming] The _S_ isn't the start of a new word:

```suggestion
    function getIntersectionTypePart(
```